caps/repo.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package caps
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+// Repository stores all known snappy capabilities and types
+type Repository struct {
+	m sync.Mutex // protects the internals from concurrent access. If contention gets high, switch to a RWMutex
+	// Map of capabilities, indexed by Capability.Name
+	caps map[string]*Capability
+	// A slice of types that are recognized and accepted
+	types []Type
+}
+
+// NewRepository creates an empty capability repository
+func NewRepository() *Repository {
+	return &Repository{
+		caps:  make(map[string]*Capability),
+		types: make([]Type, 0),
+	}
+}
+
+// Add a capability to the repository.
+// Capability names must be valid snap names, as defined by ValidateName, and
+// must be unique within the repository.  An error is returned if this
+// constraint is violated.
+func (r *Repository) Add(cap *Capability) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject capabilities with invalid names
+	if err := ValidateName(cap.Name); err != nil {
+		return err
+	}
+	// Reject capabilities with duplicate names
+	if _, ok := r.caps[cap.Name]; ok {
+		return fmt.Errorf("cannot add capability %q: name already exists", cap.Name)
+	}
+	// Reject capabilities with unknown types
+	if !r.HasType(cap.Type) {
+		return fmt.Errorf("cannot add capability %q: type %q is unknown", cap.Name, cap.Type)
+	}
+	// Reject capabilities that don't pass type-specific validation
+	if err := cap.Type.Validate(cap); err != nil {
+		return err
+	}
+	r.caps[cap.Name] = cap
+	return nil
+}
+
+// HasType checks if the repository contains the given type
+func (r *Repository) HasType(t Type) bool {
+	for _, tt := range r.types {
+		if tt == t {
+			return true
+		}
+	}
+	return false
+}
+
+// AddType adds a capability type to the repository.
+// It's an error to add the same capability type more than once.
+func (r *Repository) AddType(t Type) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	if err := ValidateName(t.String()); err != nil {
+		return err
+	}
+	for _, otherT := range r.types {
+		if t == otherT {
+			return fmt.Errorf("cannot add type %q: name already exists", t)
+		}
+	}
+	r.types = append(r.types, t)
+	return nil
+}
+
+// Remove removes the capability with the provided name.
+// Removing a capability that doesn't exist silently does nothing
```

Comment:
Oh, that's true. I must have messed up something (or I was blind and didn't update the docstring). I'll fix it.